executor/write.go

```diff
@@ -933,43 +982,194 @@ func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datu
 			var key []byte
 			var distinct bool
 			key, distinct, err = v.GenIndexKey(ctx.GetSessionVars().StmtCtx,
-				colVals, recordIDs[i], nil)
+				colVals, handles[i], nil)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
 			if !distinct {
 				continue
 			}
-			keysWithErr = append(keysWithErr, keyWithDupError{key, kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key '%s'", recordIDs[i], v.Meta().Name)})
+			keysWithErr = append(keysWithErr, keyWithDupError{
+				false,
+				key,
+				kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key '%s'", handles[i], v.Meta().Name),
+				handles[i],
+				newRowValue,
+			})
 		}
 		rowWithKeys = append(rowWithKeys, keysWithErr)
 	}
 	return rowWithKeys, nil
 }
 
-// batchMarkDupRows marks rows with duplicate errors as nil.
-// All duplicate rows were marked and appended as duplicate warnings
-// to the statement context in batch.
-func batchMarkDupRows(ctx sessionctx.Context, t table.Table, rows [][]types.Datum) ([][]types.Datum, error) {
+// batchGetInsertKeys uses batch-get to fetch all key-value pairs to be checked for ignore or duplicate key update.
+func batchGetInsertKeys(ctx sessionctx.Context, t table.Table, newRows [][]types.Datum) ([][]keyWithDupError, map[string][]byte, error) {
 	// get keys need to be checked
-	rowWithKeys, err := getKeysNeedCheck(ctx, t, rows)
+	keysInRows, err := getKeysNeedCheck(ctx, t, newRows)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
 
 	// batch get values
 	nKeys := 0
-	for _, v := range rowWithKeys {
+	for _, v := range keysInRows {
 		nKeys += len(v)
 	}
 	batchKeys := make([]kv.Key, 0, nKeys)
```

Comment:
We can simply use `len(keysInRows)` to make `batchKeys`.
The capacity doesn't matter much.

Reply:
I think it is a suggestion from someone when optimize `insert ignore`.